src/workerd/api/streams/standard.c++

```diff
@@ -2080,6 +2103,184 @@ ReadableStreamJsController::getController() {
+  // of producing either a single concatenated kj::Array<byte> or kj::String.
+public:
+  using Readable = kj::Own<T>;
+  using PartList = kj::Array<kj::ArrayPtr<byte>>;
+
+  AllReader(Readable readable, uint64_t limit): state(kj::mv(readable)), limit(limit) {
+    AllReaderBase* base = this;
+    state.template get<Readable>()->setOwner(base);
+  }
+
+  void doClose() override {
+    state.template init<StreamStates::Closed>();
+  }
+
+  void doError(jsg::Lock& js, v8::Local<v8::Value> reason) override {
+    state.template init<StreamStates::Errored>(js.v8Ref(reason));
+  }
+
+  jsg::Promise<kj::Array<byte>> allBytes(jsg::Lock& js) {
+    return loop(js).then(js, [this](auto& js, PartList&& partPtrs) {
+      auto out = kj::heapArray<byte>(runningTotal);
+      copyInto(out, kj::mv(partPtrs));
+      return kj::mv(out);
+    });
+  }
+
+  jsg::Promise<kj::String> allText(jsg::Lock& js) {
+    return loop(js).then(js, [this](auto& js, PartList&& partPtrs) {
+      auto out = kj::heapArray<char>(runningTotal + 1);
+      copyInto(out.slice(0, out.size() - 1).asBytes(), kj::mv(partPtrs));
+      out.back() = '\0';
+      return kj::String(kj::mv(out));
+    });
+  }
+private:
+  kj::OneOf<StreamStates::Closed, StreamStates::Errored, Readable> state;
+  uint64_t limit;
+  kj::Vector<jsg::BackingStore> parts;
+  uint64_t runningTotal = 0;
+
+  jsg::Promise<PartList> loop(jsg::Lock& js) {
+    KJ_SWITCH_ONEOF(state) {
+      KJ_CASE_ONEOF(closed, StreamStates::Closed) {
+        return js.resolvedPromise(KJ_MAP(p, parts) { return p.asArrayPtr(); });
+      }
+      KJ_CASE_ONEOF(errored, StreamStates::Errored) {
+        return js.template rejectedPromise<PartList>(errored.getHandle(js));
+      }
+      KJ_CASE_ONEOF(readable, Readable) {
+        const auto read = [&](auto& js) {
+          if constexpr (kj::isSameType<T, ByteReadable>()) {
+            return readable->read(js, nullptr);
+          } else {
+            return readable->read(js);
+          }
+        };
+
+        return read(js).then(js, [this](auto& js, ReadResult result) {
+          if (result.done) {
+            KJ_ASSERT(result.value == nullptr);
+            doClose();
+            return loop(js);
+          }
+
+          // If we're not done, the result value must be interpretable as
+          // bytes for the read to make any sense.
+          auto handle = KJ_ASSERT_NONNULL(result.value).getHandle(js);
+          if (!handle->IsArrayBufferView() && !handle->IsArrayBuffer()) {
+            auto error = js.v8TypeError("This ReadableStream did not return bytes.");
+            doError(js, error);
+            return loop(js);
+          }
+
+          jsg::BufferSource bufferSource(js, handle);
+          jsg::BackingStore backing = bufferSource.detach(js);
+          runningTotal += backing.size();
+          if (runningTotal >= limit) {
+            auto error = js.v8TypeError("Memory limit exceeded before EOF.");
```

Comment:
Couldn't `runningTotal == limit` at this point, but the next `read()` return with `result.done`? That is, should the condition above be `>`, not `>=`? This might matter if, for instance, we're expecting to read a length-prefixed message, we read the length, then read the payload with a limit equal to the expected length.

Reply:
Actually, we should just move this check up before the `runningTotal` is incremented and make it a `if (runningTotal + backing.size() > limit)` check...